File: batchwin/netfeed.py
#   Copyright 2013 Free Software Foundation, Inc.
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.

#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.

import re
import struct
import btqwopts
import gbclient
import netmsg

from PyQt4.QtCore import *
from PyQt4.QtGui import *
from PyQt4.QtNetwork import *

charmatch = re.compile("[\x00-\x08\x0b-\x1f\x7f-\xff]")

def net_feed(job):
    """Initialise socket to provide required feed and return it"""
    feedpkt = netmsg.feeder(job_number = job.bj_job)
    socket = QTcpSocket()
    socket.connectToHost(job.netid, btqwopts.Options.ports.jobview)
    if not socket.waitForConnected(1000):
        socket.close()
        return None
    gbclient.pushwrite(socket, feedpkt.encode())
    return socket

def replesc(gr):
    """Substitute sequences for control characters"""
    ch = ord(gr.group(0))
    result = ""
    if ch == 13:
        return ""
    if ch >= 128:
        result = "M-"
        ch -= 128
    if ch == 127: return result + "DEL"
    if ch < 32:
        result += '^'
        ch += 64
    return result + chr(ch)

def feed_string(socket):
    """Get a bufferworth of data"""
    socket.waitForReadyRead(500)
    buf = socket.read(1024)
    if not buf:
        return ""
    return charmatch.sub(replesc, buf)

File: src/lib/net_feed.c
/* net_feed.c -- Get myself a spool file from remote machine

   Copyright 2009 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include <stdio.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#ifdef  HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include "defaults.h"
#include "incl_net.h"
#include "incl_unix.h"
#include "files.h"
#include "btmode.h"
#include "btconst.h"
#include "timecon.h"
#include "bjparam.h"
#include "btjob.h"
#include "btvar.h"
#include "netmsg.h"
#include "q_shm.h"

/* Grab hold of a spool file on a remote system.  */

FILE *net_feed(const int type, const netid_t netid, const jobno_t jobno, const int port)
{
        int                     sock;
        FILE                    *result;
        struct  feeder          fd;
        struct  sockaddr_in     sin;

        if  ((sock = socket(PF_INET, SOCK_STREAM, 0)) < 0)
                return  (FILE *) 0;

        /* Set up bits and pieces The port number is set up in the job
           shared memory segment.  */

        sin.sin_family = AF_INET;
        sin.sin_port = port;
        BLOCK_ZERO(sin.sin_zero, sizeof(sin.sin_zero));
        sin.sin_addr.s_addr = netid;

        if  (connect(sock, (struct sockaddr *) &sin, sizeof(sin)) < 0)  {
                close(sock);
                return  (FILE *) 0;
        }

        /* Send out initial packet saying what we want.  */

        fd.fdtype = (char) type;
        fd.jobno = htonl(jobno);
        if  (write(sock, (char *) &fd, sizeof(fd)) != sizeof(fd))  {
                close(sock);
                return  (FILE *) 0;
        }

        /* Create result descriptor.
           We are strictly speaking cheating saying it's read-only,
           but we don't plan to do any writing and stdio doesn't mind. */

        if  ((result = fdopen(sock, "r")) == (FILE *) 0)  {
                close(sock);
                return  (FILE *) 0;
        }
#ifdef  SETVBUF_REVERSED
        setvbuf(result, _IOFBF, (char *) 0, BUFSIZ);
#else
        setvbuf(result, (char *) 0, _IOFBF, BUFSIZ);
#endif
        return  result;
}
